src/Cli/dotnet/Commands/Run/VirtualProjectBuildingCommand.cs

```diff
@@ -128,6 +192,157 @@ public int Execute(string[] binaryLoggerArgs, ILogger consoleLogger)
+                value != otherValue)
+            {
+                Reporter.Verbose.WriteLine($"""
+                    Building because previous global property "{key}" ({otherValue}) does not match current ({value}): {successCacheFile.FullName}
+                    """);
+                return true;
+            }
+        }
+
+        DateTime buildTimeUtc = successCacheFile.LastWriteTimeUtc;
+
+        // Check that the source file is up to date.
+        // If it does not exist, we also want to build.
+        if (!entryPointFileInfo.Exists || entryPointFileInfo.LastWriteTimeUtc > buildTimeUtc)
+        {
+            Reporter.Verbose.WriteLine("Building because entry point file is missing or modified: " + entryPointFileInfo.FullName);
+            return true;
+        }
+
+        // Check that implicit build files are up to date.
+        foreach (var implicitBuildFilePath in previousCacheEntry.ImplicitBuildFiles.Keys)
+        {
+            var implicitBuildFileInfo = new FileInfo(implicitBuildFilePath);
+            if (!implicitBuildFileInfo.Exists || implicitBuildFileInfo.LastWriteTimeUtc > buildTimeUtc)
+            {
+                Reporter.Verbose.WriteLine("Building because implicit build file is missing or modified: " + implicitBuildFileInfo.FullName);
+                return true;
+            }
+        }
+
+        // Check that no new implicit build files are present.
+        foreach (var implicitBuildFilePath in cacheEntry.ImplicitBuildFiles.Keys)
+        {
+            if (!previousCacheEntry.ImplicitBuildFiles.ContainsKey(implicitBuildFilePath))
+            {
+                Reporter.Verbose.WriteLine("Building because new implicit build file is present: " + implicitBuildFilePath);
+                return true;
+            }
+        }
+
+        return false;
+
+        static RunFileBuildCacheEntry? DeserializeCacheEntry(FileInfo cacheFile)
+        {
+            try
+            {
+                using var stream = File.Open(cacheFile.FullName, FileMode.Open, FileAccess.Read, FileShare.Read);
+                return JsonSerializer.Deserialize(stream, RunFileJsonSerializerContext.Default.RunFileBuildCacheEntry);
+            }
+            catch (Exception e)
+            {
+                Reporter.Verbose.WriteLine($"Failed to deserialize cache entry ({cacheFile.FullName}): {e.GetType().FullName}: {e.Message}");
+                return null;
+            }
+        }
+    }
+
+    private void MarkBuildStart()
+    {
+        string directory = GetArtifactsPath();
+        Directory.CreateDirectory(directory);
+        File.WriteAllText(Path.Join(directory, BuildStartCacheFileName), EntryPointFileFullPath);
+    }
+
+    private void MarkBuildSuccess(RunFileBuildCacheEntry cacheEntry)
+    {
+        string successCacheFile = Path.Join(GetArtifactsPath(), BuildSuccessCacheFileName);
+        using var stream = File.Open(successCacheFile, FileMode.Create, FileAccess.Write, FileShare.Read);
```

Comment:
Do we want `FileShare.Read` here? Should we block others from reading the file until we have closed it?